src/snowflake/snowpark/mock/_functions.py

```diff
@@ -1186,7 +1186,7 @@ def mock_substring(
     base_expr: ColumnEmulator, start_expr: ColumnEmulator, length_expr: ColumnEmulator
 ):
     res = [
-        x[y - 1 : y + z - 1] if x is not None else None
+        x[y - 1 : y + z - 1] if y != 0 else x[y : y + z] if x is not None else None
         for x, y, z in zip(base_expr, start_expr, length_expr)
```

Comment:
You could simplify the nested if statement by moving the 0 check into the zip expression. 
```suggestion
        None if string is None else string[start : start + length]
        for string, start, end in zip(base_expr, [max(0, s - 1) for s in start_expr], length_expr)
```